### PR TITLE
AJ-765 - Invoke Leo to create WDS<>CBAS upon Workspace creation

### PR DESCRIPTION
**Pull request by aaronkanzer**
Relates to https://broadworkbench.atlassian.net/browse/AJ-765

Invokes Leo's createV2App endpoint if a given Terra Workspace does not have a corresponding WDS App running yet.

### CODE REVIEW

File: src/components/NewWorkspaceModal.js [hunk 1]
```diff
@@ -111,7 +111,12 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
         }],
         async () => {
           const workspace = await Ajax().Workspaces.create(body)
-          Ajax().Metrics.captureEvent(Events.workspaceCreate, extractWorkspaceDetails(
+          // Only invoke Leo if we are within an Azure Workspace
+          if (!workspace.googleProject) {
+            await Ajax().Leo.create(workspace.name, workspace.id)
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js]: (I know this is still in draft) - the appName here should be "cbas-wds-default" (which should probably move to a constant somewhere) so `getWdsUrl ` can easily find it: https://github.com/DataBiosphere/terra-ui/blob/67469562e6de0d629b6fd8d67427ba2a915df690/src/libs/ajax/data-table-providers/WdsDataTableProvider.ts#L89

Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Is `cbas-wds-default` a placeholder value for now? Or is this something we want to name all WDS apps for the foreseeable future?

File: src/components/NewWorkspaceModal.js [hunk 2]
```diff
@@ -111,7 +111,12 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
         }],
         async () => {
           const workspace = await Ajax().Workspaces.create(body)
-          Ajax().Metrics.captureEvent(Events.workspaceCreate, extractWorkspaceDetails(
+          // Only invoke Leo if we are within an Azure Workspace
+          if (!workspace.googleProject) {
```
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: Please use `isAzureWorkspace`.

File: src/components/NewWorkspaceModal.js [hunk 3]
```diff
@@ -111,7 +111,12 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
...
+            await Ajax().Leo.create(workspace.name, workspace.id)
+          }
+          // TODO: Probably should add a metric for the Leo creation too
+          await Ajax().Metrics.captureEvent(Events.workspaceCreate, extractWorkspaceDetails(
```
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: No need to add the await. This is a fire and forget sort of thing, and we usually do not await `captureEvent`.

I would suggest you remove the TODO. Either add the metric, or create a ticket for someone's backlog, but having a TODO here isn't particularly helpful.

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: I couldn't use `isAzureWorkspace` in this context, since the payload returned from `Ajax().Workspaces.create(body)` doesn't contain the `cloudPlatform` key.

I could update the function of `isAzureWorkspace`, but unfortunately as is, `!workspace.googleProject` was the only way I could understand if it was Azure. Let me know your thoughts here!

File: src/components/NewWorkspaceModal.js (hunk 3 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Removed the `TODO` comment here!

File: src/components/NewWorkspaceModal.js [hunk 4]
```diff
@@ -111,10 +111,15 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
         }],
         async () => {
           const workspace = await Ajax().Workspaces.create(body)
+          // Only invoke Leo if we are within an Azure Workspace
+          if (!workspace.googleProject) {
+            await Ajax().Leo.create('cbas-wds-default', workspace.workspaceId)
+          }
```
Comment (calypsomatic) [inline src/components/NewWorkspaceModal.js]: Have we settled on using the workspaceId (or other unique name) as the appName here?

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts [hunk 5]
```diff
@@ -84,6 +84,14 @@ const getRelationParts = (val: unknown): string[] => {
 
 // Extract wds URL from Leo response. exported for testing
 export const getWdsUrl = apps => {
+  // TODO: AJ-761: Check for if app.status === 'PROVISIONING'
```
Comment (calypsomatic) [inline src/libs/ajax/data-table-providers/WdsDataTableProvider.ts]: I think this info will be used for a different ticket than 761 - probably https://broadworkbench.atlassian.net/browse/AJ-783

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (davidangb) [reply src/components/NewWorkspaceModal.js]: the Rawls swagger docs say that `cloudPlatform` is available, though you get `{"cloudPlatform":"whatever"}` instead of the `{"workspace": {"cloudPlatform": "whatever"}}` that `isAzureWorkspace` expects. Just tossing this out there; I don't know if it is worth refactoring `isAzureWorkspace` to support this.

File: src/components/NewWorkspaceModal.js (hunk 4 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Would we want to perhaps append it? e.g. `cbas-wds-default-{workspace.workspaceId}`?

Comment (davidangb) [reply src/components/NewWorkspaceModal.js]: +1, let's incorporate the workspaceId into the appName. Maybe "wds-"+workspace.workspaceId?

File: src/components/NewWorkspaceModal.js [hunk 6]
```diff
@@ -111,10 +111,15 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
         }],
         async () => {
           const workspace = await Ajax().Workspaces.create(body)
+          // Only invoke Leo if we are within an Azure Workspace
+          if (!workspace.googleProject) {
+            await Ajax().Leo.create('cbas-wds-default', workspace.workspaceId)
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js]: do you need the `await` here? Can app creation be fire-and-forget? Or do we actually want to wait for the Leo response before doing anything else?

File: src/components/NewWorkspaceModal.js [hunk 7]
```diff
@@ -111,10 +111,15 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
         }],
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js]: the clone workspace condition here will also need a call to deploy the WDS app

File: src/components/NewWorkspaceModal.js [hunk 8]
```diff
@@ -111,10 +111,15 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
...
           const workspace = await Ajax().Workspaces.create(body)
+          // Only invoke Leo if we are within an Azure Workspace
+          if (!workspace.googleProject) {
+            await Ajax().Leo.create('cbas-wds-default', workspace.workspaceId)
+          }
           Ajax().Metrics.captureEvent(Events.workspaceCreate, extractWorkspaceDetails(
             // Create response does not include cloudPlatform.
             _.merge(workspace, { cloudPlatform: getProjectCloudPlatform() }))
           )
+          setCreating(false)
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js]: why does this need an extra `setCreating(false)`? Was this not working correctly before?

File: src/libs/ajax.js [hunk 9]
```diff
@@ -418,6 +418,17 @@ const CromIAM = signal => ({
   }
 })
 
+// TODO: AJ-717: Since WDS<>CBAS are coupled, this is just one endpoint -- eventually, this endpoint can be separate per app that Leo may deploy
```
Comment (davidangb) [inline src/libs/ajax.js]: no changes requested for this PR, but for the future: you could certainly accept `appType` as an argument and remove the hardcoded `CROMWELL` in the request body

File: src/libs/ajax.js [hunk 10]
```diff
@@ -1038,7 +1049,8 @@ export const Ajax = signal => {
     FirecloudBucket: FirecloudBucket(signal),
     OAuth2: OAuth2(signal),
     Surveys: Surveys(signal),
-    WorkspaceData: WorkspaceData(signal)
+    WorkspaceData: WorkspaceData(signal),
+    Leo: Leo(signal)
```
Comment (davidangb) [inline src/libs/ajax.js]: instead of a new `Leo` object, how about putting the `create` endpoint under `Apps`? That would make it parallel to the `getV2AppInfo` endpoint. You'd probably want to rename `create` to `createAppV2`.

https://github.com/DataBiosphere/terra-ui/blob/5971897227d85cacb9b247cc23057dc9c86a5d61/src/libs/ajax/Apps.js#L64

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Here is an example payload that is returned when calling `https://rawls.dsde-dev.broadinstitute.org/api/workspaces` there is no `cloudPlatform` value in it :/

```
{"attributes":{"description":"aaron-new-test"},"authorizationDomain":[],"bucketName":"","createdBy":"aaronkanzer@gmail.com","createdDate":"2023-01-18T19:44:58.243Z","googleProject":"","isLocked":false,"lastModified":"2023-01-18T19:44:58.243Z","name":"new-test","namespace":"ajSharedBilling2023","workspaceId":"e941c71e-9cb0-4d9d-9649-7e91ef9fe25a","workspaceType":"mc","workspaceVersion":"v2"}
```

Could we include `!workspace.googleProject` as an `||` logic in `isAzureWorkspace`?

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts [hunk 11]
```diff
@@ -84,6 +84,14 @@ const getRelationParts = (val: unknown): string[] => {
...
+  // An assumption is made that there is only one app, can we make this assumption?
+  const provisioningApp = apps.filter(app => app.status === 'PROVISIONING')
+  if (provisioningApp.length === 1) {
+    // app deployment still in progress
+    return 'PROVISIONING'
+  }
```
Comment (davidangb) [inline src/libs/ajax/data-table-providers/WdsDataTableProvider.ts]: I suggest leaving this - and the associated changes in Data.js - out of the PR. We can tackle those enhancements in a future PR; I don't want them to block the auto-deployment which is critical.

File: src/components/NewWorkspaceModal.js (hunk 1 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Updated alongside changes for https://github.com/DataBiosphere/terra-ui/pull/3673#discussion_r1073943051

File: src/components/NewWorkspaceModal.js (hunk 4 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Updated here! Let me know if what I have there now is 👍

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts (hunk 5 above)
Comment (aaronkanzer) [reply src/libs/ajax/data-table-providers/WdsDataTableProvider.ts]: Updated!

File: src/components/NewWorkspaceModal.js (hunk 7 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Do you mean here another call to Leo here to make an identical set of pods? Or rather another WDS instance?

File: src/components/NewWorkspaceModal.js (hunk 6 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Good catch -- removed!

File: src/components/NewWorkspaceModal.js (hunk 8 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Ah forgot to remove -- was an artifact of some testing there -- removed!

File: src/libs/ajax.js (hunk 10 above)
Comment (aaronkanzer) [reply src/libs/ajax.js]: Good point -- refactored here to go with your suggestion in `Apps.js`

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts (hunk 11 above)
Comment (aaronkanzer) [reply src/libs/ajax/data-table-providers/WdsDataTableProvider.ts]: Removed!

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts [hunk 12]
```diff
@@ -86,7 +86,7 @@ const getRelationParts = (val: unknown): string[] => {
 export const getWdsUrl = apps => {
   // look explicitly for an app named 'cbas-wds-default'. If found, use it, even if it isn't running
```
Comment (calypsomatic) [inline src/libs/ajax/data-table-providers/WdsDataTableProvider.ts]: Nit: This comment is wrong now

**Review by calypsomatic — APPROVED** [review]

File: src/components/NewWorkspaceModal.js (hunk 4 above)
Comment (jordansafer) [reply src/components/NewWorkspaceModal.js]: Edit: ignore, I see you already made the change :)
We also may want to update the logic for getting the proxy url to look for the new wds name. Currently it looks for the appName 'cbas-wds-default'. I think this is the only other reference to appName.

**Review by jordansafer — APPROVED** [review]

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts (hunk 12 above)
Comment (aaronkanzer) [reply src/libs/ajax/data-table-providers/WdsDataTableProvider.ts]: Updated!

**Review by jordansafer — APPROVED** [review]

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (davidangb) [reply src/components/NewWorkspaceModal.js]: ah interesting! So, the Rawls swagger definition is just plain wrong then - blech.

File: src/components/NewWorkspaceModal.js (hunk 7 above)
Comment (davidangb) [reply src/components/NewWorkspaceModal.js]: I mean another call to Leo.

There are two ways for an end user to make a new workspace: create one from scratch, or clone a pre-existing workspace. In both cases, the result is a new workspace … however, given the code in this PR a user will only get WDS if they create from scratch, not clone. So this PR needs to ensure that it deploys WDS in both cases.

File: src/components/NewWorkspaceModal.js (hunk 4 above)
Comment (davidangb) [reply src/components/NewWorkspaceModal.js]: After seeing all the comments I like your name suggestion of `cbas-wds-default-{workspace.workspaceId}` better than mine :) But any of those names will work, as long as they include the workspaceId.

**Review by davidangb — CHANGES_REQUESTED** [review]
looks good, with the exception of needing to handle the clone-workspace case (especially since cloning a featured workspace is the main use case for release!)

File: src/components/NewWorkspaceModal.js (hunk 7 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: ah that makes sense -- I've added a similar API call there as well now.

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: Ugh, I forgot that create and clone workspace don't include `cloudPlatform` (which I actually commented on at line 122!). I will create a Workspaces ticket for us to add that to the response. In the meantime, define a variable with the result of `getProjectCloudPlatform()`, and use the variable here and at line 122.

Created https://broadworkbench.atlassian.net/browse/WOR-723

File: src/components/NewWorkspaceModal.js [hunk 13]
```diff
@@ -99,6 +99,9 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
       onSuccess(await Utils.cond(
         [cloneWorkspace, async () => {
           const workspace = await Ajax().Workspaces.workspace(cloneWorkspace.workspace.namespace, cloneWorkspace.workspace.name).clone(body)
+          if (!workspace.googleProject) {
```
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: Please create a variable for the cloudPlatform using `cloneWorkspace.workspace.cloudPlatform` (include the comment at line 109), and use it both here and at line 110.

This will get cleaned up as part of https://broadworkbench.atlassian.net/browse/WOR-723.

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: Actually, you can use `isAzureBillingProject` here, and leave line 122 alone.

File: src/components/NewWorkspaceModal.js (hunk 13 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: Or for a more straightforward change, use `isAzureBillingProject` here and leave line 110 alone.

Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Updated!

File: src/components/NewWorkspaceModal.js (hunk 4 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Souinds good -- I've updated the naming convention here!

File: src/components/NewWorkspaceModal.js (hunk 2 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Updated!

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts [hunk 14]
```diff
@@ -84,9 +84,9 @@ const getRelationParts = (val: unknown): string[] => {
 
 // Extract wds URL from Leo response. exported for testing
 export const getWdsUrl = apps => {
-  // look explicitly for an app named 'cbas-wds-default'. If found, use it, even if it isn't running
+  // look explicitly for an app named 'wds-${app.workspaceId}'. If found, use it, even if it isn't running
   // this handles the case where the user has explicitly shut down the app
-  const namedApp = apps.filter(app => app.appType === 'CROMWELL' && app.appName === 'cbas-wds-default')
+  const namedApp = apps.filter(app => app.appType === 'CROMWELL' && app.appName === `wds-${app.workspaceId}`)
```
Comment (cahrens) [inline src/libs/ajax/data-table-providers/WdsDataTableProvider.ts:89]: I find it confusing that the appType is `CROMWELL`, but the `appName` is wds (which I thought was a different app). Can you explain?

File: src/components/NewWorkspaceModal.js [hunk 15]
```diff
@@ -99,18 +99,25 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
       onSuccess(await Utils.cond(
         [cloneWorkspace, async () => {
           const workspace = await Ajax().Workspaces.workspace(cloneWorkspace.workspace.namespace, cloneWorkspace.workspace.name).clone(body)
+          if (isAzureBillingProject()) {
```
Comment (cahrens) [inline src/components/NewWorkspaceModal.js]: Could you please move the event firing code up above the call to `createAppV2`, just in case the Leo call fails?

I also wonder what the experience will be if the workspace creates, but the Leo call fails. Hopefully it will be clear what happened.

Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: Actually, could you move this `createApp` call out of here and do it in a separate await (with its own error handling) after the workspace creation and event firing has happened? That way you don't have to duplicate the call in the 2 locations, and the error messaging could be more specific.

File: src/libs/ajax/data-table-providers/WdsDataTableProvider.ts (hunk 14 above)
Comment (aaronkanzer) [reply src/libs/ajax/data-table-providers/WdsDataTableProvider.ts:89]: `CROMWELL` is an enumeration required by Leo (which requires an `appType`) to invoke the correct Helm charts for deploying WDS, etc.: See here: https://leonardo.dsde-dev.broadinstitute.org/#/appsV2/createAppV2

The enumeration is eventually returned when we query Leo for running WDS apps

File: src/components/NewWorkspaceModal.js (hunk 15 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: Sorry I'm not quite sure I follow -- where exactly would you want to move the call?

Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: I am thinking that that the code in the `onSuccess` callback needs to look like this:

1) Instantiate the new workspace (through the conditional code that either clones an existing workspace or does does a create, plus its metrics event firing).

2) After the previous step has been executed, create the Leo app for Azure workspaces.

In other words, the code for creating the Leo app should only be in 1 place, instead of being inserted into both conditions in step 1.

As far as error handling, I'm not clear if the call to Leo needs to have its own error handling, or if things will be reasonable if we allow any errors to fall into the existing catch block. I recommend that you try doing something that will error (for example, just change the URL being hit to be wrong) and see what happens.

File: src/components/NewWorkspaceModal.js [hunk 16]
```diff
@@ -99,18 +99,25 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
       onSuccess(await Utils.cond(
         [cloneWorkspace, async () => {
           const workspace = await Ajax().Workspaces.workspace(cloneWorkspace.workspace.namespace, cloneWorkspace.workspace.name).clone(body)
+          if (isAzureBillingProject() && !getConfig().isProd) {
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js]: doesn't `isAzureBillingProject` require a project passed in as an argument?

File: src/components/NewWorkspaceModal.js [hunk 17]
```diff
@@ -99,18 +99,25 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
...
+            Ajax().Apps.createAppV2(`wds-${workspace.workspaceId}`, workspace.workspaceId)
+          }
           const featuredList = await Ajax().FirecloudBucket.getFeaturedWorkspaces()
           Ajax().Metrics.captureEvent(Events.workspaceClone, {
             featured: _.some({ namespace: cloneWorkspace.workspace.namespace, name: cloneWorkspace.workspace.name }, featuredList),
             ...extractCrossWorkspaceDetails(cloneWorkspace, {
               // Clone response does not include cloudPlatform, cross-cloud cloning is not supported.
-              workspace: _.merge(workspace, { cloudPlatform: cloneWorkspace.workspace.cloudPlatform })
+              workspace: _.merge(workspace, { cloudPlatform: getProjectCloudPlatform() })
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js:115]: I know there's been some back-and-forth on this code … as it stands, does this line actually need to change? This seems unrelated to the auto-deploy code.

File: src/components/NewWorkspaceModal.js (hunk 16 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: No it does not (see implementation above, which gets the current project if a project is not passed in).

File: src/components/NewWorkspaceModal.js (hunk 17 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js:115]: It does not need to change, but there is also not harm in changing it.

Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js:115]: I've kept it as is with the proposed code -- happy to change back too though!

File: src/components/NewWorkspaceModal.js (hunk 15 above)
Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js]: I added some refactoring (creating the `createLeoApp` helper function), and then call it once thereafter. Let me know if this is alright with what you had in mind!

Comment (cahrens) [reply src/components/NewWorkspaceModal.js]: I pushed up a change to the error handling. I want the source of the error to be clear if the Leo ajax call errors _after_ the workspace creation already has succeeded. With what I pushed up, this is what you will see (and you can expand for details):

image

With what you had, you are stuck with the New Workspace modal left open and spinning, and at least when the error I mocked (just changing the URL), you didn't see the reason for the failure. 

Since the workspace itself has created and will be returned in the list of workspaces, we want to make sure that it is clear where the error occurred.

Note… you may wish a different error summary message than "Error starting Leo app".

Comment (davidangb) [reply src/components/NewWorkspaceModal.js]: pretty sure the marketing term is "Workspace Services," so a good summary is "Error starting Workspace Services." You might wanna get Beth's input on that, though

File: src/components/NewWorkspaceModal.js [hunk 18]
```diff
@@ -83,6 +83,12 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
     setIsAlphaRegionalityUser(await Ajax(signal).Groups.group(getConfig().alphaRegionalityGroup).isMember())
   })
 
+  const createLeoApp = withErrorIgnoring(async workspace => {
```
Comment (cahrens) [inline src/components/NewWorkspaceModal.js:88]: It might be good to add a comment as to why you are ignoring the error.

Comment (aaronkanzer) [reply src/components/NewWorkspaceModal.js:88]: added!

**Review by cahrens — APPROVED** [review]

File: src/components/NewWorkspaceModal.js [hunk 19]
```diff
@@ -83,6 +83,14 @@ const NewWorkspaceModal = withDisplayName('NewWorkspaceModal', ({
     setIsAlphaRegionalityUser(await Ajax(signal).Groups.group(getConfig().alphaRegionalityGroup).isMember())
   })
 
+  // Error is ignored here since any errors associated with Leo app creation are not communicated during workspace creation
+  // Rather when the user first visits the `Data` tab
+  const createLeoApp = withErrorIgnoring(async workspace => {
+    if (isAzureBillingProject() && !getConfig().isProd) {
+      await Ajax().Apps.createAppV2(`wds-${workspace.workspaceId}`, workspace.workspaceId)
+    }
+  })
```
Comment (davidangb) [inline src/components/NewWorkspaceModal.js:92]: I _think_ that you could remove the async/await here:
```javascript
  const createLeoApp = withErrorIgnoring(workspace => {
    if (isAzureBillingProject() && !getConfig().isProd) {
      Ajax().Apps.createAppV2(`wds-${workspace.workspaceId}`, workspace.workspaceId)
    }
  })
```
BUT, since this ignores errors and nothing relies on the result of `createLeoApp`, the change to remove async/await is invisible to users; it just makes slightly more work for the JS engine. I'm fine leaving this as-is.

**Review by davidangb — APPROVED** [review]
:+1:, one optional comment inline which is not a blocker. Thanks for getting this done!

File: src/components/NewWorkspaceModal.js (hunk 19 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js:92]: We aren't awaiting `createLeoApp`, so the creation of the workspace proceeds before this finishes. But the async/await here is necessary to make sure that the error handling happens.

(See https://github.com/DataBiosphere/terra-ui/pull/3699#discussion_r1081889331)

**Review by Qi77Qi — COMMENTED** [review]
1. will it make sense to give user an option to whether creating WDS when creating workspace? (like I test jupyter all the time, but don't really look at WDS/cromwell much)
2. Does this increase workspace creation time? or it's async to workspace creation?

**Review by n8landolt — APPROVED** [review]
Changes to Apps.js looks good!

**Comment by aaronkanzer** [comment]
> 1. will it make sense to give user an option to whether creating WDS when creating workspace? (like I test jupyter all the time, but don't really look at WDS/cromwell much)
> 2. Does this increase workspace creation time? or it's async to workspace creation?

For 1 -- since it takes a few minutes to launch WDS, @bethsheets / Product decided that launching WDS upon Workspace Creation (Azure workspaces only) would be preferable. 

For 2 -- it is async to workspace creation, so no delay in time :D

File: src/components/NewWorkspaceModal.js (hunk 17 above)
Comment (nawatts) [reply src/components/NewWorkspaceModal.js:115]: Check if the case of billing project cloud platform matches that expected by workspace functions. Workspace endpoints return `Gcp` and `Azure` while billing projects return `GCP` and `AZURE`.

File: src/components/NewWorkspaceModal.js (hunk 19 above)
Comment (davidangb) [reply src/components/NewWorkspaceModal.js:92]: ah - `withErrorIgnoring` needs the async/await, that makes sense, thank you I didn't realize that!

**Comment by davidangb** [comment]
> will it make sense to give user an option to whether creating WDS when creating workspace? 

Yes, but as a future enhancement, and as a more generalized feature far out of scope for this PR … choose your features when creating a workspace, add/remove features once it's been created.

File: src/components/NewWorkspaceModal.js (hunk 17 above)
Comment (cahrens) [reply src/components/NewWorkspaceModal.js:115]: That's a good point, but for the purpose of the event emitting it doesn't matter-- I'm always upper-casing the cloudPlatform for consistency.

https://github.com/DataBiosphere/terra-ui/blob/3718f12d3ef210d2cf1b0935992c71cb8efb1f90/src/libs/events.js#L123

Comment (nawatts) [reply src/components/NewWorkspaceModal.js:115]: Agreed the case doesn't matter for metrics, as long as it's consistent.
